dcos_test_utils/dcos_cli.py

```diff
@@ -0,0 +1,136 @@
+import logging
+import os
+import shutil
+import stat
+import subprocess
+import tempfile
+
+import requests
+
+log = logging.getLogger(__name__)
+
+DCOS_CLI_URL = "https://downloads.dcos.io/binaries/cli/linux/x86-64/latest/dcos"
+
+
+class DcosCli():
+
+    def __init__(self, cli_path):
+        self.path = os.path.abspath(os.path.expanduser(cli_path))
+        updated_env = os.environ.copy()
+        # make sure the designated CLI in on to of the PATH
+        updated_env.update({
+            'PATH': "{}:{}".format(
+                os.path.dirname(self.path),
+                os.environ['PATH']),
+            'PYTHONIOENCODING': 'utf-8',
+            'PYTHONUNBUFFERED': 'x'
+        })
+        self.env = updated_env
+
+    @classmethod
+    def new_cli(cls, download_url=DCOS_CLI_URL):
+        tmpdir = tempfile.mkdtemp()
+        dcos_cli_path = os.path.join(tmpdir, "dcos")
+        requests.packages.urllib3.disable_warnings()
+        with open(dcos_cli_path, 'wb') as f:
+            r = requests.get(download_url, stream=True, verify=True)
+            for chunk in r.iter_content(1024):
+                f.write(chunk)
+
+        # make binary executable
+        st = os.stat(dcos_cli_path)
+        os.chmod(dcos_cli_path, st.st_mode | stat.S_IEXEC)
+
+        return cls(dcos_cli_path)
+
+    @staticmethod
+    def clear_cli_dir():
+        shutil.rmtree(os.path.expanduser("~/.dcos"))
+
+    def exec_command(self, cmd, stdin=None):
+        """Execute CLI command and processes result.
+
+        This method expects that process won't block.
+
+        :param cmd: Program and arguments
+        :type cmd: [str]
+        :param stdin: File to use for stdin
+        :type stdin: file
+        :returns: A tuple with stdout and stderr
+        :rtype: (str, str)
+        """
+
+        log.info('CMD: {!r}'.format(cmd))
+
+        process = subprocess.run(
+            cmd,
+            stdin=stdin,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=self.env,
+            check=True)
+
+        stdout, stderr = process.stdout.decode('utf-8'), process.stderr.decode('utf-8')
+
+        log.info('STDOUT: {}'.format(stdout))
+        log.info('STDERR: {}'.format(stderr))
+
+        return (stdout, stderr)
+
+    def setup_enterprise(self, url, username=None, password=None):
```

Comment:
Would it make more sense to have an enterprise cli class that inherits from this one and then just keeps the username and password as instance variables? Would you ever want to setup or login as different users on the same cli object?

Reply:
I dont think we need to introduce a new interface for sub-command CLIs... The CLI is very stateful on its own. Honestly, I don't even think this should be in this class at all (tests should be the one setting this up) but I didnt want to rock to boat too much.

All that being said, it becomes a pattern to attach special helpers to these CLI abstractions then yeah, we might want that inherited class